cabal-testsuite/cabal-testsuite.cabal

```diff
@@ -126,6 +126,8 @@ executable test-runtime-deps
   build-depends:
     , Cabal
     , Cabal-syntax
+    , cabal-install
+    , cabal-install-solver
```

Comment:
These extra dependencies could be the reason why CI fails for old GHCs. https://github.com/haskell/cabal/actions/runs/10374787522/job/28728475366?pr=8889#step:10:18

Reply:
That's what we figured on Matrix too. But is there a way out of it without removing the deps? That said, I've not looked in how important it is to depend on these. It feels a little suspicious to me that these deps had to be added. Maybe, some stuff from these packages should be moved into Cabal/Cabal-syntax?  Needs more research to tell positively.

Reply:
Use `build-tool-depends: cabal-install, cabal-install-solver` instead, maybe?

Reply:
This doesn't answer, though, why the solver considers only cabal-install-3.12.1.0 and earlier but not 3.13, does it?..

Also doesn't explain why it fails for old GHCs only. For the record, the setup we use there is tricky because there are two GHCs at play: one that builds cabal (typically, GHC_FOR_RELEASE version) and one that's being driven by cabal (an old GHC like 8.0, 8.2, etc.).

Reply:
`build-tool-depends` needs to reference an executable, so I could add `cabal-install` via `build-tool-depends: cabal-install:cabal`. Unfortunately `cabal-install-solver` is not an executable, so I can not add it via `build-tool-depends` :(

About the reason to  add both dependencies:
The tests I added use some types of both packages in the assertions to verify the read ProjectConfig contains correct values, e.g.
- from cabal-install: `Distribution.Client.BuildReports.Types (ReportLevel (..)), Distribution.Client.CmdInstall.ClientInstallFlags (ClientInstallFlags (..))` and more
- from cabal-install-solver the following types: `Distribution.Solver.Types.ConstraintSource (ConstraintSource (..)),  Distribution.Solver.Types.ProjectConfigPath (ProjectConfigPath (..)) and Distribution.Solver.Types.Settings`

The cabal-testsuite setup is certainly tricky, but I will continue to try to debug it.

Reply:
cabal-testsuite was meant as an integration-test suite, I believe. That is, tests in it are supposed to call the cabal binary. If you want to write something closer to unit tests relying on cabal-install API, you may be better off with adding these tests to a test component in the cabal-install package proper. How does that sound?

Reply:
Ah, I thought all new tests should come into the cabal-testsuite. Yes, a test component in cabal-install package would be more fitting here.
I guess I could take inspiration from `Cabal-tests/tests/ParserTests.hs`. 

What do you think about adding a new test-suite to cabal-install? Something similar to the following:
```cabal
test-suite parser-tests
  type:             exitcode-stdio-1.0
  hs-source-dirs:   tests
  main-is:          ParserTests.hs
...
```

Given that I already plan to introduce something similar to the existing WarningTests in a followup issue, I could add them there, too. But for now, I would just migrate the tests there, should not be too much work.

Reply:
A new component sounds fine. 

We want to eventually move all tests out of cabal-install, actually, which should make the life of certain downstream packagers easier. But I think for this patch it's fine to still use the same package. Someone else will have to investigate a problem like we have here at another time. This is a completely separate concern from this PR.

Reply:
I moved the tests into a separate component to `cabal-install/parser-tests` and added executing it to `validate.sh`. CI is all green now :)

Reply:
Hooray!